### PR TITLE
Support multiple WithDeathAnimation traits

### PR DESCRIPTION
**Pull request by abcdefg30**
We need that if we want different palettes for different death types.

### CODE REVIEW

**Comment by GraionDilach** [comment]
Incase anyone wonders why these changes.

Our die3 was unused in TS - and the animation is a headshot at 90% of the units with only hijacker and cyborg commando excluded - so it's better if we refer that as such.
Especially considering that the actual ExplosionDeath used s_bang34.
EnergyDeath is the RA1 tesla death animation, the only difference is being the format converted from RA1 SHP to TS SHP. TS was shipped with that animation using wrong palette.

This will mean that we enable a unused death animation inserted before TS die3, die4 and die5.

**Comment by Mailaender** [comment]
Looking good. :+1: Infantry still dies properly in Tiberian Sun. :white_check_mark:

File: OpenRA.Mods.Common/Traits/Render/WithDeathAnimation.cs [hunk 1]
```diff
@@ -80,11 +80,15 @@ public void Killed(Actor self, AttackInfo e)
 			if (Info.UseDeathTypeSuffix)
 			{
 				var warhead = e.Warhead as DamageWarhead;
-				var damageType = warhead.DamageTypes.Intersect(Info.DeathTypes.Keys).FirstOrDefault();
-				if (damageType == null)
+				var wda = self.TraitsImplementing<WithDeathAnimation>().FirstOrDefault(s => s.Info.DeathTypes.Keys.Any(k => warhead.DamageTypes.Contains(k)));
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithDeathAnimation.cs]: This feels really ugly, and is now covered by a lint check.  How about we just remove it?

File: OpenRA.Mods.Common/Traits/Crushable.cs [hunk 2]
```diff
@@ -49,7 +49,7 @@ public void WarnCrush(Actor crusher)
 		public void OnCrush(Actor crusher)
 		{
 			Sound.Play(info.CrushSound, crusher.CenterPosition);
-			var wda = self.TraitOrDefault<WithDeathAnimation>();
+			var wda = self.TraitsImplementing<WithDeathAnimation>().FirstOrDefault(s => s.Info.CrushedSequence != null);
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Crushable.cs]: Style nit: Could you please put the `.FirstOrDefault` on a new line, indented by one additional level?

**Comment by abcdefg30** [comment]
Updated.

File: OpenRA.Mods.Common/Traits/Crushable.cs (hunk 2 above)
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Crushable.cs]: I've noticed that you like linebreaks, but that's not long enough to need one. :/

File: mods/ts/rules/defaults.yaml [hunk 3]
```diff
@@ -227,13 +227,21 @@
...
 	RenderSprites:
 	WithInfantryBody:
-	WithDeathAnimation:
+	WithDeathAnimation@normal:
 		DeathTypes:
 			BulletDeath: 1
 			SmallExplosionDeath: 2
-			ExplosionDeath: 3
-			InfantryExplodeDeath: 4 #not used by any warhead; TODO: check what should use this type and add it!
+			HeadshotDeath: 3 #not used by any warhead; TODO: check what should use this type and add it!
```
Comment (penev92) [inline mods/ts/rules/defaults.yaml]: Something is wrong here.
You are removing the currently unused death type, but keeping the "This is not used!" comment on the used type.

**Comment by penev92** [comment]
You are removing the `throw` that is there as a safety measure to ensure this doesn't break silently, so please make sure the lint check for this does not need changing. Would be pretty bad if we break the lint check.

File: mods/ts/rules/defaults.yaml (hunk 3 above)
Comment (GraionDilach) [inline mods/ts/rules/defaults.yaml]: No, something is getting fixed there. Our die3 did not existed in TS and before this comment, our die4 had no sequence.

TS die3 was InfantryExplode=S_BANG34. (See your tibsun.mix/local.mix/rules.ini along with http://modenc.renegadeprojects.com/InfDeath). When @reaperrr managed to sneak in said unused (infdie.shp) asset into the game, he inserted it before the last 3 used deathanims (explosion, flameguy, electro)

infdie is a headshot animation. What he did there is to restore the used death type and rename the unused one. ExplosionDeath got shifted from 3 to 4 and the other got a new name. The to-do still stands.

**Comment by abcdefg30** [comment]
Lint still works.

File: mods/ts/rules/defaults.yaml (hunk 3 above)
Comment (penev92) [inline mods/ts/rules/defaults.yaml]: OK, can we replace "TODO: check what should use this type and add it!" with "TODO: should be used by/for X and Y"?

Comment (GraionDilach) [inline mods/ts/rules/defaults.yaml]: Umagon comes into my mind as a use-case.

Comment (abcdefg30) [inline mods/ts/rules/defaults.yaml]: I think I should do that in a follow-up PR then.

**Comment by penev92** [comment]
:+1:

**Comment by penev92** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/f3f5bdd6f80d3edf77bfdaa967e680cf969bc837...75546a0ff8d02e60ba1c4bd3a24acad33f5af217)

**Pull request merged**